File: app/src/main/java/com/example/myapplication/Main/Searching.java

package com.example.myapplication.Main;

import android.content.Intent;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Build;
import android.os.Bundle;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;

import java.text.SimpleDateFormat;
import java.time.LocalDate;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import com.example.myapplication.Database.Activity;
import com.example.myapplication.Database.User;
import com.example.myapplication.R;

public class Searching extends AppCompatActivity {

    List<Activity> list;
    FirebaseDatabase firebaseDatabase;
    DatabaseReference databaseReference;
    FirebaseStorage firebaseStorage;
    RecyclerViewAdapter myAdapter;

    Date startDateFilterDate;
    Date endDateFilterDate;
    SimpleDateFormat sdf;
    Date date;
    LocationManager mLocationManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_searching);
        getSupportActionBar().hide();

        list = new ArrayList<>();

        FirebaseAuth mAuth = FirebaseAuth.getInstance();
        firebaseDatabase = FirebaseDatabase.getInstance("https://activity-1f1ae-default-rtdb.europe-west1.firebasedatabase.app");
        databaseReference = firebaseDatabase.getReference(Activity.class.getSimpleName());
        firebaseStorage = FirebaseStorage.getInstance();
        //String activityName = getIntent().getStringExtra("activity");
        String activityName = "jogging";
        String sortBy = "distance";
        int ageMinFilter = 8;
        int ageMaxFilter = 30;
        int distanceFilter = 1000000;
        sdf = new SimpleDateFormat("dd-MM-yyyy");
        String startDateFilter = "23-09-2021";
        String endDateFilter = "30-09-2021";
        String sexFilter = "Male";
        double latitude = getIntent().getDoubleExtra("latitude", 0);
        double longitude = getIntent().getDoubleExtra("longitude",0);
        Location myLocation = new Location("");
        myLocation.setLatitude(latitude);
        myLocation.setLongitude(longitude);
        Date currentDate; // to check if activity current
        try {
            startDateFilterDate = sdf.parse(startDateFilter);
            endDateFilterDate = sdf.parse(endDateFilter);
        }
        catch (Exception e) {

        }

        Query query = databaseReference.orderByChild("activity").equalTo(activityName);
        query.addChildEventListener(new ChildEventListener() {
            @RequiresApi(api = Build.VERSION_CODES.O)
            @Override
            public void onChildAdded(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {
                Activity activity = dataSnapshot.getValue(Activity.class);
                try {
                    date = sdf.parse(activity.getDate());
                }
                catch (Exception e) {
                    activity.setName(e.toString());
                }

                // here check if current outdated
                double distance = distance(myLocation.getLatitude(),myLocation.getLongitude(),activity.getLatitude(),activity.getLongitude());
                if (activity.getActivity().equals(activityName)
                        && ageMinFilter<activity.getAge() && activity.getAge()<ageMaxFilter
                        && (startDateFilterDate.before(date) ||startDateFilterDate.equals(date))
                        && (endDateFilterDate.after(date) ||endDateFilterDate.equals(date))
                        && (sexFilter.equals(activity.getSex()) || activity.getSex() == null)
                        && (distance < distanceFilter)
                        && (activity.getUserId() != mAuth.getCurrentUser().getUid())) {

                    LocalDate localDate = LocalDate.now();
                    ZoneId zoneId = ZoneId.systemDefault();
                    ZonedDateTime zonedDateTime = localDate.atStartOfDay(zoneId);
                    Date utilDate = Date.from(zonedDateTime.toInstant());

                    Calendar a = getCalendar(date);
                    Calendar b = getCalendar(utilDate);
                    long daysBetween = ChronoUnit.DAYS.between(a.toInstant(), b.toInstant());

                    activity.setDaysTo(daysBetween);

                    activity.setDistance(Math.round(distance/100)/10.0); //Math.round(distance/100)/10
                    list.add(activity);
                }

                if(sortBy == "date") {
                    Collections.sort(list, new Comparator<Activity>() {
                        @Override
                        public int compare(Activity a1, Activity a2) {
                            Date d1 = null;
                            Date d2 = null;
                            try {
                                d1 = sdf.parse(a1.getDate());
                                d2 = sdf.parse(a2.getDate());
                            }
                            catch (Exception e) {
                            }
                            return d1.compareTo(d2);
                        }
                    });
                }
                else if(sortBy == "distance") {
                    Collections.sort(list, new Comparator<Activity>() {
                        @Override
                        public int compare(Activity a1, Activity a2) {
                            if (distance(myLocation.getLatitude(),myLocation.getLongitude(),a1.getLatitude(),a1.getLongitude())
                                    > distance(myLocation.getLatitude(),myLocation.getLongitude(),a2.getLatitude(),a2.getLongitude())) {
                                return 1;
                            }
                            else return -1;
                        }
                    });
                }
                myAdapter.notifyDataSetChanged();
            }

            @Override
            public void onChildChanged(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {

            }

            @Override
            public void onChildRemoved(@NonNull DataSnapshot dataSnapshot) {

            }

            @Override
            public void onChildMoved(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {

            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });

        RecyclerView myrv = (RecyclerView) findViewById(R.id.recyclerview_id);
        myAdapter = new RecyclerViewAdapter(this, list);
        myrv.setLayoutManager(new GridLayoutManager(this,2));
        myrv.setAdapter(myAdapter);

    }

    private Location getLastKnownLocation() {
        mLocationManager = (LocationManager)getApplicationContext().getSystemService(LOCATION_SERVICE);
        List<String> providers = mLocationManager.getProviders(true);
        Location bestLocation = null;
        for (String provider : providers) {
            Location l = mLocationManager.getLastKnownLocation(provider);
            if (l == null) {
                continue;
            }
            if (bestLocation == null || l.getAccuracy() < bestLocation.getAccuracy()) {
                // Found best last known location: %s", l);
                bestLocation = l;
            }
        }
        return bestLocation;
    }

    private double distance(double lat1, double lon1, double lat2, double lon2) {
        double theta = lon1 - lon2;
        double dist = Math.sin(deg2rad(lat1)) * Math.sin(deg2rad(lat2)) + Math.cos(deg2rad(lat1)) * Math.cos(deg2rad(lat2)) * Math.cos(deg2rad(theta));
        dist = Math.acos(dist);
        dist = rad2deg(dist);
        dist = dist * 60 * 1.1515* 1.609344;
        return (dist);
    }

    /*:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::*/
    /*::  This function converts decimal degrees to radians             :*/
    /*:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::*/
    private double deg2rad(double deg) {
        return (deg * Math.PI / 180.0);
    }

    /*:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::*/
    /*::  This function converts radians to decimal degrees             :*/
    /*:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::*/
    private double rad2deg(double rad) {
        return (rad * 180.0 / Math.PI);
    }

    public static Calendar getCalendar(Date date) {
        Calendar cal = Calendar.getInstance(Locale.US);
        cal.setTime(date);
        return cal;
    }

}

File: app/src/main/java/com/example/myapplication/main/Searching.kt
package com.example.myapplication.main

import android.Manifest
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationManager
import android.os.Build
import android.os.Bundle
import android.view.View
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.myapplication.database.Activity
import com.example.myapplication.R
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import com.google.firebase.storage.FirebaseStorage
import java.text.SimpleDateFormat
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.ChronoUnit
import java.util.*
import kotlin.math.acos
import kotlin.math.cos
import kotlin.math.roundToInt
import kotlin.math.sin

class Searching : AppCompatActivity() {
    val list: MutableList<Activity> = ArrayList()
    private var firebaseDatabase: FirebaseDatabase? = null
    private var databaseReference: DatabaseReference? = null
    private var firebaseStorage: FirebaseStorage? = null
    var myAdapter: RecyclerViewAdapter? = null
    var startDateFilterDate: Date? = null
    var endDateFilterDate: Date? = null
    var sdf: SimpleDateFormat? = null
    var date: Date? = null
    private var mLocationManager: LocationManager? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_searching)
        supportActionBar!!.hide()
        val mAuth = FirebaseAuth.getInstance()
        firebaseDatabase =
            FirebaseDatabase.getInstance("https://activity-1f1ae-default-rtdb.europe-west1.firebasedatabase.app")
        databaseReference = firebaseDatabase!!.getReference(Activity::class.java.simpleName)
        firebaseStorage = FirebaseStorage.getInstance()
        //String activityName = getIntent().getStringExtra("activity");
        val activityName = "jogging"
        val sortBy = "distance"
        val ageMinFilter = 8
        val ageMaxFilter = 30
        val distanceFilter = 1000000
        sdf = SimpleDateFormat("dd-MM-yyyy", Locale.US)
        val startDateFilter = "23-09-2021"
        val endDateFilter = "30-09-2021"
        val sexFilter = "Male"
        val latitude = intent.getDoubleExtra("latitude", 0.0)
        val longitude = intent.getDoubleExtra("longitude", 0.0)
        val myLocation = Location("")
        myLocation.latitude = latitude
        myLocation.longitude = longitude

        startDateFilterDate = sdf!!.parse(startDateFilter)
        endDateFilterDate = sdf!!.parse(endDateFilter)

        val query = databaseReference!!.orderByChild("activity").equalTo(activityName)
        query.addChildEventListener(object : ChildEventListener {
            @RequiresApi(api = Build.VERSION_CODES.O)
            override fun onChildAdded(dataSnapshot: DataSnapshot, s: String?) {
                val activity = dataSnapshot.getValue(
                    Activity::class.java
                )
                try {
                    date = sdf!!.parse(activity!!.date!!)
                } catch (e: Exception) {
                    activity!!.name = e.toString()
                }

                // here check if current outdated
                val distance = distance(
                    myLocation.latitude,
                    myLocation.longitude,
                    activity!!.latitude,
                    activity.longitude
                )
                if (activity.activity == activityName &&
                    ageMinFilter < activity.age &&
                    activity.age < ageMaxFilter &&
                    (startDateFilterDate!!.before(date) || startDateFilterDate == date) &&
                    (endDateFilterDate!!.after(date) || endDateFilterDate == date) &&
                    (sexFilter == activity.sex || activity.sex == null) &&
                    distance < distanceFilter &&
                    activity.userId !== mAuth.currentUser!!.uid
                ) {
                    val localDate = LocalDate.now()
                    val zoneId = ZoneId.systemDefault()
                    val zonedDateTime = localDate.atStartOfDay(zoneId)
                    val utilDate = Date.from(zonedDateTime.toInstant())
                    val a = getCalendar(date)
                    val b = getCalendar(utilDate)
                    val daysBetween = ChronoUnit.DAYS.between(a.toInstant(), b.toInstant())
                    activity.daysTo = daysBetween
                    activity.distance = (distance / 100).roundToInt() / 10.0 //Math.round(distance/100)/10
                    list.add(activity)
                }
                if (sortBy === "date") {
                    list.sortWith{ a1, a2 ->
                        val d1 = sdf!!.parse(a1.date!!)
                        val d2 = sdf!!.parse(a2.date!!)

                        d1!!.compareTo(d2)
                    }
                } else if (sortBy === "distance") {
                    list.sortWith{ a1, a2 ->
                        if (distance(
                                myLocation.latitude,
                                myLocation.longitude,
                                a1.latitude,
                                a1.longitude
                            )
                            > distance(
                                myLocation.latitude,
                                myLocation.longitude,
                                a2.latitude,
                                a2.longitude
                            )
                        ) {
                            1
                        } else -1
                    }
                }
                myAdapter!!.notifyDataSetChanged()
            }

            override fun onChildChanged(dataSnapshot: DataSnapshot, s: String?) {}
            override fun onChildRemoved(dataSnapshot: DataSnapshot) {}
            override fun onChildMoved(dataSnapshot: DataSnapshot, s: String?) {}
            override fun onCancelled(databaseError: DatabaseError) {}
        })
        val myRv = findViewById<View>(R.id.recyclerview_id) as RecyclerView
        myAdapter = RecyclerViewAdapter(this, list)
        myRv.layoutManager = GridLayoutManager(this, 2)
        myRv.adapter = myAdapter
    }

    // Found best last known location: %s", l);
    private val lastKnownLocation: Location?
        get() {
            mLocationManager =
                applicationContext.getSystemService(LOCATION_SERVICE) as LocationManager
            val providers = mLocationManager!!.getProviders(true)
            var bestLocation: Location? = null
            for (provider in providers) {
                if (ActivityCompat.checkSelfPermission(
                        this,
                        Manifest.permission.ACCESS_FINE_LOCATION
                    ) == PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
                        this,
                        Manifest.permission.ACCESS_COARSE_LOCATION
                    ) == PackageManager.PERMISSION_GRANTED
                ) {
                    val l = mLocationManager!!.getLastKnownLocation(provider) ?: continue

                    if (bestLocation == null || l.accuracy < bestLocation.accuracy) {
                        // Found best last known location: %s", l);
                        bestLocation = l
                    }
                }
            }
            return bestLocation
        }

    private fun distance(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val theta = lon1 - lon2
        var dist =
            sin(deg2rad(lat1)) * sin(deg2rad(lat2)) + cos(deg2rad(lat1)) * cos(
                deg2rad(lat2)
            ) * cos(deg2rad(theta))
        dist = acos(dist)
        dist = rad2deg(dist)
        dist *= 60 * 1.1515 * 1.609344
        return dist
    }

    /*:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::*/ /*::  This function converts decimal degrees to radians             :*/ /*:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::*/
    private fun deg2rad(deg: Double): Double {
        return deg * Math.PI / 180.0
    }

    /*:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::*/ /*::  This function converts radians to decimal degrees             :*/ /*:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::*/
    private fun rad2deg(rad: Double): Double {
        return rad * 180.0 / Math.PI
    }

    companion object {
        fun getCalendar(date: Date?): Calendar {
            val cal = Calendar.getInstance(Locale.US)
            cal.time = date!!
            return cal
        }
    }
}